### PR TITLE
Manage whitelisted guilds and file formats in DB. #305 

### LINKED ISSUE #305: Manage whitelisted guilds and file formats in DB.
This is part 2 of https://github.com/python-discord/bot/issues/639

In order to manage whitelist items via a command, we will need to store them in our database.

- Create models for the whitelist items
- Create a migration that adds these models, ~~and also automatically adds the currently whitelisted items, which you can find in `default-config.yml` on the bot repo. We don't wanna have to add these manually.~~ Let's just add the items manually, so that we can merge the site-side changes independently of the bot changes. It'll take like, 2 minutes.
- Lastly, ~~create API endpoints with all CRUD actions.~~ let Django Restful Framework handle all the CRUD actions.

### PR DESCRIPTION
**Pull request by atmishra**
This PR closes #305. This adds API to manage whitelist in the Database. Includes following changes:

1. Add Model for Whitelist
2. Add viewset, serializer for the API
3. Add migration to create existing whitelist

### CODE REVIEW

File: pydis_site/apps/api/urls.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
     DocumentationLinkViewSet, InfractionViewSet,
     LogEntryViewSet, NominationViewSet,
     OffTopicChannelNameViewSet, ReminderViewSet,
-    RoleViewSet, TagViewSet, UserViewSet
+    RoleViewSet, TagViewSet, UserViewSet, WhitelistViewSet
```
Comment (jchristgit) [inline pydis_site/apps/api/urls.py]: the other lines are all 2 items per line, can you move `UserViewSet, WhitelistViewSet` on a new line?

File: pydis_site/apps/api/models/bot/whitelist.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+from pydis_site.apps.api.models.utils import ModelReprMixin
+
+
+class Whitelist(ModelReprMixin, models.Model):
+    """Whitelisted item for a type."""
+
+    INVITE = "invite"
+    EXTENSION = "extension"
+    CHANNEL = "channel"
+    ROLE = "role"
```
Comment (jchristgit) [inline pydis_site/apps/api/models/bot/whitelist.py]: hmm, what's the reason for these constants? I don't see them used anywhere else, but if they should be, I think a separate Enum would be a better fit (e.g. `WhitelistType`)

**Review by jchristgit — CHANGES_REQUESTED** [review]
Sorry for the late review. Can you

- Resolve the merge conflicts
- Rename the  `0050_auto_20191117_1821.py` migration to something more meaningful, e.g. `0050_import_whitelisted_items`

the icing on the cake would be if the data import migration would only delete the imported items, but I don't think that's necessary

**Comment by atmishra** [comment]
Thanks! for the review. I will work on the changes.

File: pydis_site/apps/api/models/bot/whitelist.py (hunk 2 above)
Comment (atmishra) [reply pydis_site/apps/api/models/bot/whitelist.py]: Removed the constants.

Comment (jchristgit) [reply pydis_site/apps/api/models/bot/whitelist.py]: Hmm, did you forget to push, or am I looking at old code here? Constants are still here.

Comment (atmishra) [reply pydis_site/apps/api/models/bot/whitelist.py]: My Bad! I removed the use of the constants but forgot to remove them.

**Review by jchristgit — APPROVED** [review]
Code-wise this looks great!

Shortly before a merge, we need to update the whitelist to match what's currently in the bot's configuration file, since the data migration and the configuration file have drifted apart a bit since the PR was opened. I think that should wait until the second review though (or with a second data migration after merge).

**Comment by lemonsaurus** [comment]
This is basically ready for merge, but like jchrist pointed out, we should wait to merge it until the bot-side changes are ready, so that we don't have write another migration.

So, this is blocked until we've done the bot-side stuff.

The issue for the bot changes can be found in https://github.com/python-discord/bot/issues/639

File: pydis_site/apps/api/migrations/0053_merge_20200114_1559.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('api', '0049_deletedmessage_attachments'),
+        ('api', '0052_add_whitelist_unique_constraint'),
+    ]
+
+    operations = [
+    ]
```
Comment (MarkKoz) [inline pydis_site/apps/api/migrations/0053_merge_20200114_1559.py:14]: What is the purpose of this migration if it has no operations?

File: pydis_site/apps/api/models/bot/whitelist.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+
+class Whitelist(ModelReprMixin, models.Model):
+    """Whitelisted item for a type."""
+
+    TYPE_CHOICES = (
+        ("invite", "invite"),
+        ("extension", "extension"),
+        ("channel", "channel"),
+        ("role", "role")
```
Comment (MarkKoz) [inline pydis_site/apps/api/models/bot/whitelist.py:13]: Capitalise the second element of each tuple - those are the human-readable names.
```suggestion
        ("invite", "Invite"),
        ("extension", "Extension"),
        ("channel", "Channel"),
        ("role", "Role")
```
Not sure if this requires a migration.

File: pydis_site/apps/api/models/bot/whitelist.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    )
+    type = models.CharField(
+        choices=TYPE_CHOICES,
+        max_length=100,
+        help_text=(
+            "Type of the whitelisted item."
+        ),
```
Comment (MarkKoz) [inline pydis_site/apps/api/models/bot/whitelist.py:20]:
```suggestion
        help_text="Type of the whitelisted item.",
```

File: pydis_site/apps/api/models/bot/whitelist.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    )
+    whitelisted_item = models.CharField(
+        max_length=300,
+        help_text=(
+            "whitelisted item"
+        )
```
Comment (MarkKoz) [inline pydis_site/apps/api/models/bot/whitelist.py:26]:
```suggestion
        help_text="Whitelisted item."
```

File: pydis_site/apps/api/models/bot/whitelist.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    )
+    type = models.CharField(
+        choices=TYPE_CHOICES,
+        max_length=100,
+        help_text=(
+            "Type of the whitelisted item."
+        ),
+    )
+    whitelisted_item = models.CharField(
+        max_length=300,
```
Comment (MarkKoz) [inline pydis_site/apps/api/models/bot/whitelist.py:23]: This should be a `TextField` without any sort of length restriction.

File: pydis_site/apps/api/models/bot/whitelist.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+    TYPE_CHOICES = (
+        ("invite", "invite"),
+        ("extension", "extension"),
+        ("channel", "channel"),
+        ("role", "role")
+    )
+    type = models.CharField(
+        choices=TYPE_CHOICES,
+        max_length=100,
```
Comment (MarkKoz) [inline pydis_site/apps/api/models/bot/whitelist.py:17]: This should be a `TextField` without any sort of length restriction.

File: pydis_site/apps/api/viewsets/bot/whitelist.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+    mixins.CreateModelMixin,
+    mixins.ListModelMixin,
+    mixins.DestroyModelMixin,
+    GenericViewSet,
+):
+    """
+    View providing create, list and  delete operations on whitelist shown by our bot.
+
+    ## Routes
+    ### GET /bot/whitelist?type=invites
```
Comment (MarkKoz) [inline pydis_site/apps/api/viewsets/bot/whitelist.py:19]: Don't include a query parameter as part of the route. Document the query parameters in a separate section.

File: pydis_site/apps/api/viewsets/bot/whitelist.py [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+
+class WhitelistViewSet(
+    mixins.CreateModelMixin,
+    mixins.ListModelMixin,
+    mixins.DestroyModelMixin,
+    GenericViewSet,
+):
+    """
+    View providing create, list and  delete operations on whitelist shown by our bot.
```
Comment (MarkKoz) [inline pydis_site/apps/api/viewsets/bot/whitelist.py:16]: This suggestion assumes you've added `RetrieveModelMixin` and thus made it support CRUD.

```suggestion
    View providing CRUD operations on the whitelist used by the Discord bot's filters.
```

File: pydis_site/apps/api/viewsets/bot/whitelist.py [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+from pydis_site.apps.api.models.bot.whitelist import Whitelist
+from pydis_site.apps.api.serializers import WhitelistSerializer
+
+
+class WhitelistViewSet(
+    mixins.CreateModelMixin,
+    mixins.ListModelMixin,
+    mixins.DestroyModelMixin,
+    GenericViewSet,
```
Comment (MarkKoz) [inline pydis_site/apps/api/viewsets/bot/whitelist.py:13]: Wouldn't hurt to add `RetrieveModelMixin` and just make this CRUD. May come in handy in the future and it's a 1 line code change (well, besides the docstring) thanks to DRF's magic.

File: pydis_site/apps/api/viewsets/bot/whitelist.py [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+    ### POST /bot/whitelist
+    Adds a single whitelist item to the database.
+
+    #### Request body
+    >>> {
+    ...     'type': str,
+    ...     'whitelisted_item': str
+    ... }
+
```
Comment (MarkKoz) [inline pydis_site/apps/api/viewsets/bot/whitelist.py:52]: There was an extra blank line.
```suggestion
```

File: pydis_site/apps/api/models/bot/whitelist.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
+from django.db import models
+
+from pydis_site.apps.api.models.utils import ModelReprMixin
+
+
+class Whitelist(ModelReprMixin, models.Model):
+    """Whitelisted item for a type."""
```
Comment (MarkKoz) [inline pydis_site/apps/api/models/bot/whitelist.py:7]:
```suggestion
    """An item of a certain type which is excluded from the Discord bot's filters."""
```

File: pydis_site/apps/api/migrations/0051_import_whitelisted_items.py [hunk 14]
```diff
@@ -0,0 +1,77 @@
+# Generated by Django 2.2.6 on 2019-11-17 18:21
+
+from django.db import migrations
+
+
+class Migration(migrations.Migration):
+    def create_whitelist(app, schema_editor):
+        Whitelist = app.get_model("api", "Whitelist")
+        invites_whitelist = [
```
Comment (MarkKoz) [inline pydis_site/apps/api/migrations/0051_import_whitelisted_items.py:9]: Let's update this migration with the latest configuration from the bot before this PR is merged.

**Review by MarkKoz — CHANGES_REQUESTED** [review]

File: pydis_site/apps/api/models/bot/whitelist.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+    TYPE_CHOICES = (
+        ("invite", "invite"),
+        ("extension", "extension"),
```
Comment (MarkKoz) [inline pydis_site/apps/api/models/bot/whitelist.py:11]: This should be renamed to more clearly express that it's a file extension and not, say, a Discord extension. My suggestion: `("file_ext", "File Extension"),`

**Review by MarkKoz — CHANGES_REQUESTED** [review]

**Comment by atmishra** [comment]
I am working on the bot side changes which will possible to some changes in this PR. I will incorporate the review changes with the API changes.

**Comment by MarkKoz** [comment]
@atmishra what is the status of this?

File: pydis_site/apps/api/migrations/0053_merge_20200114_1559.py (hunk 3 above)
Comment (lemonsaurus) [reply pydis_site/apps/api/migrations/0053_merge_20200114_1559.py:14]: To merge two heads, and restore linear order to the migration history.

**Review by lemonsaurus — COMMENTED** [review]
This has been open without any changes for long enough.

**Pull request closed without merging**